Support NestedPlanExpression in slotted runtime

henriknyman opened the pull request:
- Interleave slot allocation for expressions with slot allocation for
logical plans.
- Recursively allocate slots for NestedPlanExpressions and merge
the information into the physical plan.
- Make sure we use the expression converter that does
NestedPlanExpression -> NestedPipeExpression everywhere in
SlottedPipeBuilder.
- Add NestedPlanExpression to SlottedRewriter and pass in the
full slot configurations map so that expressions can be
correctly rewritten within nested plans.

henriknyman committed:
Support NestedPlanExpression in slotted runtime

- Interleave slot allocation for expressions with slot allocation for
logical plans.
- Recursively allocate slots for NestedPlanExpressions and merge
the information into the physical plan.
- Make sure we use the expression converter that does
NestedPlanExpression -> NestedPipeExpression everywhere in
SlottedPipeBuilder.
- Add NestedPlanExpression to SlottedRewriter and pass in the
full slot configurations map so that expressions can be
correctly rewritten within nested plans.

henriknyman committed:
Move Morsel runtime to an experimental test configuration

This test configuration is not executed unless explicitly requested
in a test.

henriknyman committed:
Update tests that succeed with slotted runtime

henriknyman committed:
Add unit test

henriknyman committed:
Add debug pretty printing to slotted plan builder

Uses Kiama and ansi colors to show some useful debug information

systay commented on enterprise/cypher/physical-planning/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/SlotAllocation.scala:
```diff
@@ -122,11 +126,16 @@ object SlotAllocation {
           val lhsSlots = resultStack.pop()
-          val result = allocate(current, nullable, lhsSlots, rhsSlots)
-          allocations += (current.assignedId -> result)
+          // NOTE: This assumes expressions on this logical plan are evaluated in the scope of the LHS
```
Have you considered ValueHashJoin? It will evaluate expressions on both sides

systay replied:
See: https://github.com/neo4j/neo4j/blob/3.4/enterprise/cypher/physical-planning/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/SlottedRewriter.scala#L95

henriknyman replied:
Good point!

fickludd commented on enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/CypherComparisonSupport.scala:
```diff
@@ -635,13 +638,25 @@ object CypherComparisonSupport {
+      * I.e. there will be no check to see if they unexpectedly succeed on tests where they were not explicitly requested.
+      */
+    def Experimental: TestConfiguration =
+      TestConfiguration(Versions.Default, Planners.Default, Runtimes(Runtimes.Morsel))
```
This is a good way to keep `Morsel` in a maintainable way.

fickludd commented on enterprise/cypher/cypher/pom.xml:
```diff
@@ -348,6 +348,11 @@
+    <dependency>
+      <groupId>org.bitbucket.inkytonik.kiama</groupId>
+      <artifactId>kiama_2.11</artifactId>
+      <version>2.1.0</version>
```
License problem?

fickludd commented on enterprise/cypher/physical-planning/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/SlotAllocation.scala:
```diff
@@ -55,17 +55,19 @@ object SlotAllocation {
     val planStack = new mutable.Stack[(Boolean, LogicalPlan)]()
     val resultStack = new mutable.Stack[SlotConfiguration]()
     val argumentStack = new mutable.Stack[SlotsAndArgument]()
+    initialSlotsAndArgument.foreach(argumentStack.push)
```
We could use `SlotAllocation.NO_ARGUMENT()` to encode no argument and not need the `Option`. This is also ok though.

fickludd commented on enterprise/cypher/physical-planning/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/SlotAllocation.scala:
```diff
@@ -97,17 +99,19 @@ object SlotAllocation {
-          allocations += (current.assignedId -> result)
-          resultStack.push(result)
+          val result = allocateExpressions(current, nullable, sourceSlots, allocations, arguments)
+          val result2 = allocate(current, nullable, result, recordArgument(_, argument))
```
How about calling them `nestedResult` and `result` instead?

henriknyman committed:
Add LGPL 3.0 to licensing requirements

- Update licenses for new dependency
- Add a synonym to Mozilla Public License 2.0

fickludd commented on enterprise/cypher/physical-planning/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/SlotAllocation.scala:
```diff
@@ -135,6 +144,43 @@ object SlotAllocation {
+        slots => {
+          e.introducedVariables.foreach {
+            case v@Variable(name) =>
+              slots.newReference(name, true, CTAny)
```
I'm still not convinced that we should allocate local variables among the streamed slots, but if you think it's the way to go then let's try.

henriknyman committed:
Make expression slot allocation handle two sourced plans

ValueHashJoin has one expression that is evaluated on the lhs
and one that is evaluated on the rhs. We now do two passes of
allocate expression on logical plans with two sources, one for
lhs and one for rhs (and block traversal of the expression on the side
we are not currently looking at).

fickludd commented on enterprise/cypher/physical-planning/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/SlotAllocation.scala:
```diff
@@ -135,6 +144,43 @@ object SlotAllocation {
+          slotConfigurations ++= nestedPhysicalPlan.slotConfigurations
+          argumentSizes ++= nestedPhysicalPlan.argumentSizes
+
+          (slots, None)
```
But this means that `slots` is unchanged by nested plans, or am I misreading?

henriknyman committed:
Remove exception for expressions now supported by slotted runtime

fickludd commented on enterprise/cypher/physical-planning/src/test/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/SlotAllocationTest.scala:
```diff
@@ -736,4 +736,26 @@ class SlotAllocationTest extends CypherFunSuite with LogicalPlanningTestSupport2
+    allocations(nestedPlan.assignedId) should equal(
+      SlotConfiguration(Map("x" -> LongSlot(0, nullable = false, CTNode)), 1, 0)
+    )
+  }
```
Nested plans sound complicated to me... could we test something more to specify the slot configuration of 
 * upstream operators
 * the passing of the parent operator scope to the nested plan
 * the passing of the parent argument to the nested plan

henriknyman replied:
There are tests for this in the acceptance spec suite, like `CaseExpression.feature`

fickludd reviewed (changes requested):
I know you are still working on this, but before we merge this I would like some more tests around the nested plan allocation to specify the behavior in some more complicated cases. 

The implementation looks good to me.

henriknyman committed:
Remove LGPLv3 license

Exclude transitive dependencies requiring LGPLv3

henriknyman committed:
Turn off Morsel from experimental test configuration

henriknyman committed:
Turn off all Morsel test configurations

henriknyman commented:
@fickludd We discussed introducing a runtime acceptance level of testing, where we give logical plans, build and execute, and assert on the result. I think it will be suitable for the additional test coverage you wanted to see.
